flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/writer/topic/TopicProducerRegister.java

```diff
@@ -0,0 +1,176 @@
+    }
+
+    /**
+     * Get the cached topic related transaction. Or create a new transaction after checkpointing.
+     */
+    public Transaction getOrCreateTransaction(String topic) {
+        return transactionRegister.computeIfAbsent(
+                topic,
+                t -> {
+                    long timeoutMillis = sinkConfiguration.getTransactionTimeoutMillis();
+                    return createTransaction(pulsarClient, timeoutMillis);
+                });
+    }
+
+    /** Abort the existed transactions. This method would be used when close PulsarWriter. */
+    private void abortTransactions() {
+        if (transactionRegister.isEmpty()) {
+            return;
+        }
+
+        TransactionCoordinatorClient coordinatorClient =
+                ((PulsarClientImpl) pulsarClient).getTcClient();
+        // This null check is used for making sure transaction is enabled in client.
+        checkNotNull(coordinatorClient);
+
+        try (Closer closer = Closer.create()) {
+            for (Transaction transaction : transactionRegister.values()) {
+                TxnID txnID = transaction.getTxnID();
+                closer.register(() -> coordinatorClient.abort(txnID));
+            }
+
+            clearTransactions();
+        } catch (IOException e) {
+            throw new FlinkRuntimeException(e);
+        }
+    }
+
+    /** Clean these transactions. All transactions should be passed to Pulsar committer. */
+    private void clearTransactions() {
+        // Clear the transactions, we would create new transaction when new message comes.
+        transactionRegister.clear();
+    }
+
+    /**
+     * Convert the transactions into committable list for Pulsar Committer. The transactions would
+     * be removed until flink triggering a checkpoint.
+     */
+    public List<PulsarCommittable> prepareCommit() {
+        List<PulsarCommittable> committables = new ArrayList<>(transactionRegister.size());
+        transactionRegister.forEach(
+                (topic, transaction) -> {
+                    TxnID txnID = transaction.getTxnID();
+                    PulsarCommittable committable = new PulsarCommittable(txnID, topic);
+                    committables.add(committable);
+                });
+
+        clearTransactions();
+        return committables;
+    }
+
+    /**
+     * Flush all the messages buffered in the client and wait until all messages have been
+     * successfully persisted.
+     */
+    public void flush() throws IOException {
+        for (Producer<?> producer : producerRegister.values()) {
+            producer.flush();
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        try (Closer closer = Closer.create()) {
+            // Flush all the pending messages to Pulsar. This wouldn't cause exception.
+            closer.register(this::flush);
+
+            // Abort all the existed transactions.
+            closer.register(this::abortTransactions);
```

Comment:
What happens with transactions that the Committer is still trying to commit?

Reply:
They are totally not the same transactions. We would return all the pending transactions and clear them in `TopicProducerRegister` when executing `PulsarWriter#prepareCommit`. So the pending transactions in `TopicProducerRegister` only live if PulsarWriter and never be passed to `PulsarCommitter`.